### PR TITLE
feat: implement non-absorbed MLAAttention layer

### LINKED ISSUE #894: [Feature] Implement MLAAttention layer
## Summary

Implement `MLAAttention` as a shared Flax NNX module for DeepSeek-V2 style models (Ling-2.5-1T, etc.),using non-absorbed mode with existing `RadixAttention` + `SplitMHATokenToKVPool` infrastructure.

## Scope

- Q path: `hidden -> q_a_proj -> RMSNorm -> q_b_proj -> split(q_nope, q_rope)`
- KV path: `hidden -> kv_a_proj -> split(compressed, k_rope) -> RMSNorm -> kv_b_proj -> split(k_nope,v)`
- RoPE on q_rope and k_rope only
- Assembly: Q = concat(q_nope, q_rope'), K = concat(k_nope, k_rope')
- Output through RadixAttention + o_proj
- Split KV cache for K_dim(192) != V_dim(128)

## Non-Goals
- Absorbed mode (future optimization, ~35x KV cache reduction)
- Model-level integration (weight loading, config parsing)

## Acceptance Criteria

- [ ] Design docs
- [ ] `MLAAttention` module at `python/sgl_jax/srt/layers/mla.py`
- [ ] Unit tests (cosine similarity >= 0.99 vs fp64 reference): shape, Q path, KV path, full QKV, k_rope broadcast
- [ ] Integration test: `RadixAttention` + `SplitMHATokenToKVPool` end-to-end on TPU

### PR DESCRIPTION
**Pull request by zhengkezhou1**
## Motivation

fix #894

Implement the `MLAAttention` layer for MLA-based models (DeepSeek-V2/V3, Ling-2.5, etc.), as described in the design doc from #895.

This PR delivers the **non-absorbed mode** implementation: it fully decompresses the latent state into Q/K/V tensors during forward, then reuses the existing `RadixAttention` + `MHATokenToKVPool` infrastructure without requiring any MLA-specific attention kernel or compressed KV cache. This is a deliberate trade-off ~43x more KV cache memory vs. absorbed mode that prioritizes correctness and integration simplicity while the absorbed mode is still under development.

## Modifications

- **`srt/layers/mla.py`**: Add `MLAAttention` implementing the full MLA data flow:
  - Q path: `q_a_proj → RMSNorm → q_b_proj → split(q_nope, q_rope)`
  - KV path: `kv_a_proj → split(compressed, k_rope) → RMSNorm → kv_b_proj → split(k_nope, v)`
  - RoPE applied only to rope portions; V zero-padded to `qk_head_dim` for fused KV cache compatibility
  - Standard `RadixAttention` dispatch; V padding stripped before `o_proj`

- **`test/layers/test_mla.py`**: Unit tests + integration test
  - unit tests validating Q/K/V projection pipeline against a numpy fp32 reference (cosine similarity ≥ 0.99)

## Accuracy Tests

Layer-level numerical correctness is validated via cosine similarity (threshold ≥ 0.99) against an fp32 numpy reference implementation adapted from [BailingMoeV2_5MultiLatentAttention](https://huggingface.co/inclusionAI/Ling-2.5-1T/blob/main/modeling_bailing_moe_v2_5.py). Model-level e2e accuracy will be covered separately when MLA is wired into a model file.

## Checklist

- [ ] Please use English, otherwise it will be closed.
- [ ] The purpose of the PR, or link existing issues this PR will resolve.
- [ ] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/layers/mla.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        # --- RoPE (only on rope portions) ---
+        q_rope, k_rope = self.rotary_emb(positions, q_rope, k_rope)
+        k_rope = jnp.broadcast_to(k_rope, (k_rope.shape[0], self.num_heads, self.qk_rope_head_dim))
+        # broadcast_to produces unsharded output; reshard to match k_nope's
+        # tensor-parallel sharding so concatenation works on Explicit mesh.
+        k_rope = jax.device_put(
+            k_rope,
+            jax.sharding.NamedSharding(self.mesh, jax.sharding.PartitionSpec(None, "tensor", None)),
+        )
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/mla.py]: Should not use `device_put` api. Try to set `out_sharding` https://docs.jax.dev/en/latest/parallel.html

File: python/sgl_jax/srt/layers/mla.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                use_bias=False,
+                params_dtype=dtype,
+                kernel_axes=(None, "tensor"),
+                scope_name="q_b_proj",
+            )
+
+        # --- KV path: compression bottleneck ---
+        # In absorbed mode, kv_b_proj is not applied during forward — its
+        # weights are absorbed into Q (W_k portion) and output (W_v portion),
+        # and only the compressed state (kv_lora_rank) is cached.
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/mla.py]: Too much comments. Please keep the comments concise, place under the class comments.

File: python/sgl_jax/srt/layers/mla.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            base=int(rope_theta),
+            is_neox_style=not rope_interleave,
+            rope_scaling=rope_scaling,
+            dtype=dtype,
+        )
+
+        # --- Attention dispatch ---
+        # Absorbed mode replaces this with a dedicated call_mla() dispatch
+        # through FlashAttention, using MLATokenToKVPool and a custom Pallas
+        # kernel that fuses decompression into block-wise attention.
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/mla.py]: Same as above

File: python/sgl_jax/srt/layers/mla.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        # compressed latent state and decompresses inside the attention kernel.
+        v = jnp.pad(v, ((0, 0), (0, 0), (0, self.qk_head_dim - self.v_head_dim)))
+
+        k_rope = k_rope_raw.reshape(-1, 1, self.qk_rope_head_dim)
+
+        # --- RoPE (only on rope portions) ---
+        q_rope, k_rope = self.rotary_emb(positions, q_rope, k_rope)
+        k_rope = jnp.broadcast_to(k_rope, (k_rope.shape[0], self.num_heads, self.qk_rope_head_dim))
+        # broadcast_to produces unsharded output; reshard to match k_nope's
+        # tensor-parallel sharding so concatenation works on Explicit mesh.
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/mla.py]: Please reduce unnecessary comments and use methods such as inline assignment to simplify the process.

File: python/sgl_jax/test/layers/test_mla.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+            v_head_dim=V_HEAD_DIM,
+            mesh=self.mesh,
+            layer_id=0,
+            rope_theta=ROPE_THETA,
+            rope_interleave=True,
+            dtype=jnp.bfloat16,
+        )
+
+        total_q_tokens = sum(q_lens_list)
+        seq_lens = np.array(seq_lens_list, dtype=np.int32)
```
Comment (JamesBrianD) [inline python/sgl_jax/test/layers/test_mla.py]: This test case can be added after the MLA kernel & `ref_mla` are implemented; currently, it does not have the effect of verifying accuracy.

File: python/sgl_jax/srt/layers/mla.py (hunk 1 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/layers/mla.py]: Done. Replaced `jax.device_put` with `jnp.broadcast_to(..., out_sharding=P(None, "tensor", None))` for explicit sharding mode.

File: python/sgl_jax/srt/layers/mla.py (hunk 2 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/layers/mla.py]: Done. Removed verbose inline comments; kept concise info in the class docstring.

File: python/sgl_jax/srt/layers/mla.py (hunk 3 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/layers/mla.py]: Done.

Comment (zhengkezhou1) [reply python/sgl_jax/srt/layers/mla.py]: Done.

File: python/sgl_jax/srt/layers/mla.py (hunk 4 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/layers/mla.py]: Done. Simplified comments and merged broadcast + resharding into a single `jnp.broadcast_to(..., out_sharding=...)` call.

File: python/sgl_jax/test/layers/test_mla.py (hunk 5 above)
Comment (zhengkezhou1) [reply python/sgl_jax/test/layers/test_mla.py]: Done. Removed; will add back after MLA kernel & `ref_mla` are implemented.

File: python/sgl_jax/srt/layers/attention/mla.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
+"""Multi-head Latent Attention (MLA) layer."""
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/mla.py:1]: Please place this file in the attention directory.

**Review by Bob-Chen222 — DISMISSED** [review]

File: python/sgl_jax/srt/layers/attention/mla.py (hunk 6 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/layers/attention/mla.py:1]: Done. Moved mla.py to srt/layers/attention/ and updated all imports.

**Review by JamesBrianD — APPROVED** [review]
